### PR TITLE
Interaction

### LINKED ISSUE #25: Front-end interaction tracking

### PR DESCRIPTION
**Pull request by Kaleab470**
closes #25

### CODE REVIEW

File: src/main/webapp/creatives/maps/interaction.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+const URL = {
+  INITIALIZE: "initialize",
+  FIND_NEAREST_LOCATION: "find-location",
+  GRANT_LOCATION_DATA: "grant-location",
+  MAP_INTERACT: "map-interact",
+  SKIP_TO_CONTENT: "skip-to-content",
+  RETURN_TO_AD: "return-to-ad"
+}
+
+class User {
```
Comment (juliareichel) [inline src/main/webapp/creatives/maps/interaction.js:13]: I like how you structured this class I just think a better name could be used for the class, such as userSession

File: src/main/webapp/creatives/maps/utils.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+
+
+/**
+ * Sets the session ID, this should only be used on session creation.
+ */
+export function generateSessionId() {
+  let dt = new Date().getTime();
```
Comment (juliareichel) [inline src/main/webapp/creatives/maps/utils.js:7]: change variable name to something meaningful, not shorthands

File: src/main/webapp/creatives/maps/utils.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  const generateRandomHex = (c) => {
+    const r = (dt + Math.random()*16)%16 | 0;
```
Comment (juliareichel) [inline src/main/webapp/creatives/maps/utils.js:9]: change r to randomNumber

File: src/main/webapp/creatives/simid_protocol.js [hunk 4]
```diff
@@ -251,27 +252,12 @@ export default class SimidProtocol {
       // to ever communicate with the player.
       console.log('Session creation was rejected.');
     }
-    this.generateSessionId_();
+
+    this.sessionId_ = generateSessionId();
```
Comment (juliareichel) [inline src/main/webapp/creatives/simid_protocol.js:256]: does it make sense to generate an ID for a session that got rejected -- perhaps only call generateSessionId() if the session is created?

File: src/main/webapp/creatives/maps/interaction.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  constructor() {
+    this.correlator = generateSessionId();
+  }
+
+  sendRequest_(url) {
```
Comment (ryanthompson591) [inline src/main/webapp/creatives/maps/interaction.js:18]: comment this function, it's not necessarily apparent what is happening.

File: src/main/webapp/creatives/maps/interaction.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    fetch(`/${url}?correlator="${this.correlator}"`).then(() => console.log("success"));
```
Comment (ryanthompson591) [inline src/main/webapp/creatives/maps/interaction.js:19]: console.log("success") isn't a useful message.  Ideally change to a useful message or remove.

File: src/main/webapp/creatives/maps/interaction.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  initialize() {
```
Comment (ryanthompson591) [inline src/main/webapp/creatives/maps/interaction.js:22]: I would comment these functions.  Or maybe rename them.

logUserInitialized
logUserClickFindNearestLocation

or potentially if this whole class is renamed to logger or statsLogger, your use would make sense too when you call it.

For Example:
logger.userInitialized();

or
statsCollection.onUserInitialize();

way more understandable what exactly is happening compared to:
user.initialize()

File: src/main/webapp/creatives/maps/utils.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
+
+
+/**
+ * Sets the session ID, this should only be used on session creation.
```
Comment (ryanthompson591) [inline src/main/webapp/creatives/maps/utils.js:4]: This doesn't really set the session id.  You could change the definition to just have the return param

/**
 * @return {string} A statndard UUID conforming to en.wikipedia.org/wiki/Universally_unique_identifier
 */

File: src/main/webapp/creatives/maps/utils.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+ */
+export function generateSessionId() {
+  let dt = new Date().getTime();
+  const generateRandomHex = (c) => {
+    const r = (dt + Math.random()*16)%16 | 0;
+    dt = Math.floor(dt/16);
+    return (c=='r' ? r :(r&0x3|0x8)).toString(16);
+  };
+  const uuidFormat = 'rrrrrrrr-rrrr-4rrr-yrrr-rrrrrrrrrrrr';
+  const uuid = uuidFormat.replace(/[ry]/g, generateRandomHex);
```
Comment (kristen-m) [inline src/main/webapp/creatives/maps/utils.js:14]: Cool that you're using regex here!

File: src/main/webapp/creatives/maps/interaction.js (hunk 1 above)
Comment (Kaleab470) [reply src/main/webapp/creatives/maps/interaction.js]: Renamed it!

File: src/main/webapp/creatives/maps/interaction.js (hunk 7 above)
Comment (Kaleab470) [reply src/main/webapp/creatives/maps/interaction.js]: Renamed class to Logger and also renamed the methods!

File: src/main/webapp/creatives/maps/Logger.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+  SKIP_TO_CONTENT: "skip-to-content",
+  RETURN_TO_AD: "return-to-ad"
+}
+
+class Logger {
+  /**
+   * Each instance gets a correlator to uniquely identify it.
+   */
+  constructor() {
+    this.correlator = generateSessionId();
```
Comment (damargulis) [inline src/main/webapp/creatives/maps/Logger.js:19]: Give this a type, and make it private

```
/**
 * @private @const {string}
 */
 this.correlator_ = ...
```

File: src/main/webapp/creatives/maps/Logger.js [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+const URL = {
+  INITIALIZE: "initialize",
+  FIND_NEAREST_LOCATION: "find-location",
+  GRANT_LOCATION_DATA: "grant-location",
+  MAP_INTERACT: "map-interact",
+  SKIP_TO_CONTENT: "skip-to-content",
+  RETURN_TO_AD: "return-to-ad"
+}
+
+class Logger {
```
Comment (damargulis) [inline src/main/webapp/creatives/maps/Logger.js:14]: imo you can be a bit more descriptive here. `InteractionLogger` or `UserActivityLogger` or something like that

File: src/main/webapp/creatives/maps/Logger.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+  constructor() {
+    this.correlator = generateSessionId();
+  }
+
+  /**
+   * Makes a GET request to the given URL with the correlator as a parameter.
+   * @param {string} url A URL to send the request to.
+   */
+  sendRequest_(url) {
+    fetch(`/${url}?correlator="${this.correlator}"`);
```
Comment (damargulis) [inline src/main/webapp/creatives/maps/Logger.js:27]: Did you look into the reportTracking SIMID message (https://interactiveadvertisingbureau.github.io/SIMID/#simid-creative-reportTracking)

I think this is also fine -- just wondering if you had a reason not to use it

File: src/main/webapp/creatives/maps/utils.js [hunk 13]
```diff
@@ -1,7 +1,7 @@
 
 
 /**
- * Sets the session ID, this should only be used on session creation.
+ * @return {string} A statndard UUID conforming to en.wikipedia.org/wiki/Universally_unique_identifier
```
Comment (juliareichel) [inline src/main/webapp/creatives/maps/utils.js]: misspelled standard

File: src/main/webapp/creatives/maps/Logger.js (hunk 11 above)
Comment (juliareichel) [reply src/main/webapp/creatives/maps/Logger.js:14]: yes^ Logger is not any more descriptive than "User"

File: src/main/webapp/creatives/maps/utils.js (hunk 13 above)
Comment (Kaleab470) [reply src/main/webapp/creatives/maps/utils.js]: That should have been fixed in an earlier commit. Lmk if you still can't see it

**Pull request merged**